woocommerce/woocommerce-gutenberg-products-block
# Refactor block types to include script registration and integration classes

This is a fairly hefty PR which refactors our Block Type classes (and makes some changes to the Payments Integration classes).

## 1. Removes editor script registration from the Assets class to the block type classes

This keeps block type definitions together with the script registration, avoids the need to do additional checks for build type when registering assets (because blocks are not registered if they do not support the current build), and allows each block type to define it's own dependencies.

## 2. Abstracts `PaymentMethodRegistry` and `PaymentMethodTypeInterface` with `IntegrationRegistry` and `IntegrationInterface`

Since we needed something similar for other integrations (block types) I created `IntegrationRegistry` and `IntegrationInterface` and moved the code that can be reused to those.

`AbstractPaymentMethodType` takes care of the differences and this seems to be used by WC Pay etc so shouldn't break anything.

## 3. `AbstractBlock` refactor

This class will now:

- Register the editor script
- Register the frontend script
- Enqueue the frontend script when the block is rendered 
- Use the `IntegrationRegistry` based on the block name, and with this append registered dependencies and data whenever assets are registered or enqueued.

@senadir before this is merged I want to update subscriptions with it to ensure it covers all use cases. We will hook in via:

`woocommerce_blocks_{registry_identifier}_registration` is the hook where 3rd parties can register themselves. In the case of subscriptions it will be both:

- woocommerce_blocks_cart_block_registration
- woocommerce_blocks_checkout_block_registration

Fixes #3532

### How to test the changes in this Pull Request:

This touches on quite a lot of code so we need to ensure:

1. All tests pass, which should cover block rendering issues
2. Smoke test all blocks (frontend and editor) to ensure they render. This will confirm scripts are still loaded.
3. Payment methods should still show up during checkout. Test full checkout flow.
4. Check console for 404s to ensure no invalid scripts are loaded.
5. Check experimental block types like atomic blocks still function



### Changelog

> Refactor block type registration to support 3rd party integrations.

### Dev note:

An important note that internally, this release has modified how `AbstractBlock` (the base class for all of our blocks) functions, and how it loads assets. `AbstractBlock` is internal to this project and does not seem like something that would ever need to be extended by 3rd parties, but note if you are doing so for whatever reason, your implementation would need to be updated to match.
